### ISSUE TITLE
Tarkus Team Cracktro

### ISSUE THREAD

**Issue by chris70c** [dirkwhoffmann/vAmiga#522]
Severe screen corruption, missing blobs, scrolltext bug...

![vamiga](https://user-images.githubusercontent.com/3858768/132803705-a4763d9b-f61d-41c1-9dc7-7b9a53c42ab7.png)
![winuae](https://user-images.githubusercontent.com/3858768/132803710-afcd4732-5b55-4647-b989-3056bd614f73.png)

[Tarkus_Team.crack04.zip](https://github.com/dirkwhoffmann/vAmiga/files/7141468/Tarkus_Team.crack04.zip)

**Comment by dirkwhoffmann** [comment]
<img width="853" alt="Bildschirmfoto 2021-09-11 um 09 49 06" src="https://user-images.githubusercontent.com/12561945/132942301-9bcf157c-8fe0-4bb1-859b-568430c49fc1.png">

Looks like a duplicate of #518 to me. However, this one should be debugged first because the Copper list appears to be simpler. This is likely the relevant part:

<img width="377" alt="Bildschirmfoto 2021-09-11 um 09 50 31" src="https://user-images.githubusercontent.com/12561945/132942371-c1ccfccc-6cac-454d-a772-03d03de63fe8.png">

**Comment by dirkwhoffmann** [comment]
Broken with commit 6fc8b31

**Comment by dirkwhoffmann** [comment]
The bug is most likely due to a timing problem w.r.t. BPLCON0. In `Agnus::pokeBPLCON0`, if 

    recordRegisterChange(DMA_CYCLES(4), SET_BPLCON0_AGNUS, value);

is changed to 

    recordRegisterChange(DMA_CYCLES(2), SET_BPLCON0_AGNUS, value);

it works fine. 

<img width="853" alt="Bildschirmfoto 2021-09-11 um 18 16 11" src="https://user-images.githubusercontent.com/12561945/132956149-4857fd57-6bf3-47c1-9d64-19f56761cce8.png">

With a delay of 4 DMA cycles (old code), the first DMA access for bitplane 4 is skipped which causes the shift:

```
[1523] ( 68, 55) 05E120  6 BCBS-A 602C 37E0 Denise:83 setBPLCON0(3200,6400)
[1523] ( 68, 58) 05E120  6 BCBS-A 602C 37E0 Denise:213 setBPL6DAT(0)
[1523] ( 68, 59) 05E120  6 BCBS-A 602C 37E0 Denise:213 setBPL2DAT(0)
[1523] ( 68, 61) 05E120  6 BCBS-A 602C 37E0 Denise:213 setBPL3DAT(0)
[1523] ( 68, 62) 05E120  6 BCBS-A 602C 37E0 Denise:213 setBPL5DAT(0)
[1523] ( 68, 63) 05E120  6 BCBS-A 602C 37E0 Denise:213 setBPL1DAT(0)
[1523] ( 68, 65) 05E120  6 BCBS-A 602C 37E0 Denise:213 setBPL4DAT(0)
[1523] ( 68, 66) 05E120  6 BCBS-A 602C 37E0 Denise:213 setBPL6DAT(0)
```

The old  vAmiga version worked well, because the DDF window started one fetch unit later due to the missing pipeline registers in Denise.

Next step: Check if any BPLCON0 test in the test suite already covers the issue.

**Comment by dirkwhoffmann** [comment]
This one turns out to be difficult...

The good: BPLCON0 timing is OK (delay really is 4 cycles).
The bad: There is something wrong with DDF window computation.
The ugly: Fixing this requires a surgery on the open heart.

First things first: I've run three of my test cases in UAE (ddfnew1, ddfnew2, ddfnew3). These tests set DDFSTRT to $38, $3A, $3C, respectively. I let UAE print where the first bitplane DMA cycles happens (in line 69 which has been chosen arbitrarily).

    DDFSTART = 38: Bitplane 0: (69,63) 
    DDFSTART = 3A Bitplane 0: (69,63) 
    DDFSTART = 3C Bitplane 0: (69,67)

Here is the thing: Setting DDFSTRT to $3C in lores mode moves the DMA window by 4 cycles in UAE. Until now, I thought a 4 cycle shift is only possible in hires mode. I thought in lores mode, fetch units always comply to an 8 cycle raster as pictured in the HRM:

![dmaTiming](https://user-images.githubusercontent.com/12561945/133410496-c6fcbd45-5b9c-4de3-82ae-7b4b07df5e28.png)

Obviously, I have to rethink my DDF logic.

**Comment by dirkwhoffmann** [comment]
Fixed by commit 54e2419 (DDF bugfix)

**Comment by dirkwhoffmann** [comment]
Fixed in vAmiga 1.1b2
